### ISSUE TITLE
Giza: Automated Testing Plan

### ISSUE THREAD

**Issue by mnaamani** [Joystream/joystream#2590]
# GIZA Automated Testing Plan

## Coverage

- New Runtime (migration)
- New Storage system
- Updated Content Directory to work with new storage system
- New proposals related to storage system
- New Storage Node Service - Uploading assets and fetching assets
- New Storage Node CLI - lead actions
- New Distributor Node Service - Fetching replicated assets.
- New Distributor Node CLI - lead actions
- New Query Node  schema and mappings
- Content Migration Tools

## Test Environments

Basic test environment - simple so they can be easily run on developer machine as well as Github Actions
with minimal requirements (only docker without kubernetes)

### 1. Basic New Runtime - Single Machine

  - Start chain with Giza Runtime - no initial state
  - Start Query-Node
  - ~Init working groups leads and workers (content and storage)~
  - ~Start One Storage Node~
  - ~Start One Distributor Node~
  - Run standard tests* - should not make assumptions of runtime state

### 2. Basic Runtime Upgrade - Single Machine

This workflow will try to simulate as closely as possible Sumer to Giza migration scenario
(This is a separate flow because it will only need to be done once for next release)

  - Start chain with Sumer runtime - no initial state
  - Init working group lead (content and storage) and some members
  - Add some content (without upload to storage node)
  - Perform runtime upgrade to Giza
  - Test expected runtime migrations (eg. Clearing content channels/videos, storage working group?)
  - Start Query-Node (starts indexing and processing after upgrade) - test handles initial state and bootstrap correctly
  - ~Init storage working group leads and workers~
  - ~Start One Storage Node~
  - ~Start One Distributor Node~
  - Run standard tests* - should not make assumptions of runtime state

### 3. Storage Replication and Distribution
This test environment and flow will actually test the storage/distribution nodes implementations.

  - Start chain with Giza Runtime - no initial state
  - Deploy Query-Node
  - Init leads and workers (content and storage)
  - Deploy Storage Node 1
  - Deploy Distributor Node 1
  - Create channel A and some videos
       - test authentication api on storage node
       - test content policy
       - uploading content to storage node
       - verify content available on storage node
       - fetching content from distributor node
   - Delete channel and videos
       - verify storage node drops content
       - verify distributor no longer serves content?
  - Create channel B and Videos
  - Deploy new Storage Node 2
  - Deploy new Distributor Node 2
  - Update buckets so new nodes should replicate and serve content
  - Allow time for replication
  - Test content available on storage node 2
  - Test fetching content from distributor node 2

### 4. Network Test Environment - deploying network to the cloud to test storage networking

  - Deploy new network Giza Runtime - 2 validators
  - Start local rpc node
  - Deploy Query-Node
  - Init leads and workers (content and storage)
  - Deploy Storage Node 1
  - Deploy Distributor Node 1
  - Upload content with content migration tool for Giza
  - Verify all content was successfully uploaded
  - Deploy new Storage Node 2
  - Deploy new Distributor Node 2
  - Update buckets so new nodes should replicate and serve content
  - Allow time for replication
  - Test content available on storage node 2
  - Test fetching content from distributor node 2

# *Standard Tests (scenario, flows and jobs)

Standard tests are written using our testing framework and involves sending transactions, asserting events and runtime state, and any relevant query-node state by querying hydra. These tests should not require storage infrastructure (Test environments 1 and 2 above). Interaction of storage nodes with runtime can still be "mocked" as part of the tests.

In Giza the framework will need to add support for doing content uploads, and either integrating with the storage-node and distributor-node cli's (if cli packages are loadable npm modules) or executing them as external applications (former approach is more desirable)

  - Standard tests from Sumer
  - ~New Proposals related to storage~ there are none
  - Storage lead role extrinsics
  - Distributor lead role extrinsics
  - Storage Worker extrinsics
  - Distributor worker extrinsics
  - Content Directory
      - basic creation of channels and videos
         - ~uploading content to storage node~ -mocked
         - ~verify content available on storage node~
         - ~fetching content from distributor node~
      - updating channels and videos
      - deleting channels and videos
         - ~verify storage node drops content~
         - ~verify distributor no longer serves content?~
      - featured videos

# Notes

- To make the workflows simpler to write try to have test scenarios startup node services?
- Runtime work is based on Antioch.. we need to merge master branch in

**Comment by mnaamani** [comment]
## Draft tasks
Listing tasks to turn into issues under Giza project..

### Deployment/Devops
 - Update docker-compose environment with support for two storage nodes and two discovery nodes (Anuj)
 - Dockerize - Query-Node - Ensure Query-Node can be deployed with docker remotely (not using mounted local volumes) - (Anuj)
 - Dockerize Storage Node - (Shamil / Anuj)
 - Dockerize Distributor Node - (Leszek / Anuj)
 - Update Pulumi deployment scripts for all subsystems, network, query-node, new storage and discovery nodes. (Anuj)
 - Update Github Actions to run tests for all flows described in https://github.com/Joystream/joystream/issues/2590#issue-976721378 - Mokhtar/Anuj

### Tools / Scripts
- Update joystream-cli to to upload to new storage system
- Update bulk upload script to work with Giza storage system
- Update hire-lead script with sudo, and storage/distributor clis to work with non Development chain


### Testing Framework
- Add support for deploying additional nodes (storage, distributor)
- Scenario runs, ensure configurable so can work against local or remote deployment.

Can we disable storage-node MIME type filtering so we can generate random file content during upload tests.

Will we add member type dynamic bag type support (runtime integration) - if so add tests for
    - creation of bag on member creation. (Will we need to do runtime migration to create bags for existing members or should storage system create bag on demand?)

**Comment by shamil-gadelshin** [comment]
# Storage node testing plan

At the very least, all scenarios will test successful outcomes. Some scenarios could contain tests for failed attempts.

## General prerequisites
- running joystream node
- running query-node
- storage WG leader is set (this account will be used for tests)

## Scenario 1 (Single node API testing)
### Prerequisites
- uploading limits are set
- bucket is created with limits and metadata
- bag is updated
- asset is uploaded
- single storage started
### A.Files API
- get the auth-token
- upload a file
- download a file
- get file headers

### B.State API
- version
- data state
- all data objects
- data objects by bagId

## Scenario 2 (CLI testing)
### Prerequisites: command specific with no started storage nodes
- test all commands

## Scenario 3 (synchronization between two nodes)
### Prerequisites
- uploading limits are set
- bucket is created with limits and metadata
- bag is updated
- asset is uploaded
- two storage nodes started
#### A. Uploading to the first node, synchronization to the second one.
#### B. Deletion (remove bag-to-bucket association) for some node.

**Comment by Lezek123** [comment]
## Joystream CLI

### Flow: Managing curator groups via CLI

**Parameters:**
- `Tc` - total number of curators to hire
- `Tg` - total number of groups to create
- `Fg` - final number of groups
- `Ta` - total number of groups to update to status `active`
- `Fa` - final number of groups that should stay `active`
- `Tm` - total number of members to add to each group
- `Fm` - final number of members in each group

**Preconditions:**
- A running Joystream node and query-node
- Curator lead is hired

**Post-execution state:**
- There are `Fg` new curator groups, `Fa` of them are active, each group has `Fm` members

**Constraints:**
- `Tm <= Tc <= Tg * Tm`
- `0 < Fg < Tg`
- `0 < Fa < Ta`
- `0 < Fm < Tm`
- `Ta <= Tg`

**Optimistic flow:**
- Verify the output of `content:curatorGroups`: empty state
- Hire `Tc` curators (via a fixture)
- Create `Tg` curator groups
- Verify the output of `content:curatorGroups`: confirm there are `Tg` inactive, empty groups
- Verify the output `content:curatorGroup` for each group: confirm empty, inactive group
- Activate `Ta` curator groups (via `setCuratorGroupStatus`)
- Deactivate `Ta - Fa` curator groups (via `setCuratorGroupStatus`)
- Add `Tm` curators to each group
- Verify the output of `content:curatorGroups`: confirm groups statuses and member counts are valid
- Verify the output `content:curatorGroup` for each group: confirm group status and members are valid
- Remove `Tm - Fm` curators from each group
- Verify the output of `content:curatorGroups`: confirm member counts are updated
- Verify the output `content:curatorGroup` for each group: confirm members are updated
- Verify final state (`F`) against query node

**Possible additional failure cases to include:**
- Create a group using non-lead key: expect exit code `405` - `AccessDenied`
- Execute `content:curatorGroup` with invalid group id: expect exit code `400` - `InvalidInput`
- Add curator to a group using non-lead key: expect exit code `405` - `AccessDenied`
- Add cutator to non-existing group: expect exit code `400` - `InvalidInput`
- Add non-existing curator to a group: expect exit code `400` - `InvalidInput`
- Remove curator from a group using non-lead key: expect exit code `405` - `AccessDenied`
- Remove a curator from a group it doesn't belong to: expect exit code `400` - `InvalidInput`




### Flow: Channel and video categories in CLI

**Parameters:**
- `Tv` - total number of video categories to create
- `Tc` - total number of channel categories to create
- `Uv` - total number of video categories to update
- `Uc` - total number of channel categories to update
- `Fv` - final number of video categories
- `Fc` - final number of channel categories

**Constraints:**
`0 < Fc <= Tc - 2` (at least 2 channel category deletions need to be executed - one by lead and one by curator) 
`0 < Fv <= Tv - 2` (at least 2 video category deletions need to be executed - one by lead and one by curator)
`2 <= Uc <= Tc`
`2 <= Uv <= Tv`

**Preconditions:**
- A running Joystream node and query-node
- Curator lead is hired
- Optionally: There is a curator group with at least 1 curator (it can also be created in this flow)

**Post-execution state:**
- There are `Fv` new video categories and `Fc` new channel categories

**Optimistic flow:**
- Create `Tv` video categories (use curator/lead context interchangeably)
- Create `Tc` channel categories (use curator/lead context interchangeably)
- Update `Uv` video categories (use curator/lead context interchangeably)
- Update `Uc` channel categories (use curator/lead context interchangeably)
- Remove `Tv - Fv` video categories (use curator/lead context interchangeably)
- Remove `Tc - Fc` channel categories (use curator/lead context interchangeably)
- Verify final state (`F`) against query node

**Possible additional failure cases to include:**
- Create/Update/Remove channel/video category using curator context&key while not beeing part of curator group: expect exit code `405` - `AccessDenied`
- Create/Update/Remove channel/video category using non-curator key: expect exit code `405` - `AccessDenied`
- Create/Update category with invalid metadata as input: expect exit code `400` - `InvalidInput`
- Update/Remove non-existing channel/video category: expect exit code `400` - `InvalidInput`


### Flow: Channels and videos in CLI

**Preconditions:**
- A running Joystream node and query-node
- Curator lead is hired
- Optionally: There is a curator group with at least 1 curator (it can also be created in this flow)
- Optionally: Some video/channel categories are available (they can also be created in this flow)

**Optimistic flow:**
- Verify the output of `content:channels`: empty state
- Verify the output of `content:videos`: empty state
- Create member channels (try to cover all metadata cases listed below)
- Create curator group channel(s)
- Verify the output of `content:channels`: confirm channels created with expected owners, reward accs etc.
- Verify the output of `content:channel` for each channel: confirm channel data is valid in each case
- Verify channels metadata and assets data using query-node
- Update member channel(s) (try to cover all metadata cases listed below)
- Update curator channels
- Verify the output of `content:channels`: confirm channels reward accs updated where expected
- Verify the output of `content:channel` for each channel: confirm channel data updated as expected
- Verify channels metadata and assets data using query-node: confirm updated as expected
- Add some videos to each channel (test all createVideo metadata cases)
- Verify the output of `content:channels`: confirm video counts match
- Verify the output of `content:channel` for each channel: confirm video counts match
- Verify the output of `videos` - confirm all videos are listed and the data matches the expectations
- Verify the output of `videos [CHANNEL_ID]` for each channel - confirm only the expected videos are listed
- Verify the output of `video` for each video - confirm video data matches the expectations
- Verify videos metadata and assets data using query-node
- **_TBC..._**

**`createChannel` metadata cases to test (some can be combined):**
  - w/ both cover and avatar photo
  - w/ cover photo, but w/o avatar photo
  - w/ avatar photo, but w/o cover photo
  - w/ reward account set
  - with valid/invalid category set
  - with valid/invalid language set
  - w/o any metadata fields and assets set

**`updateChannel` metadata cases to test (some can be combined):**
  - setting/unsetting reward account
  - setting valid/invalid category
  - setting valid/invalid language
  - unsetting category
  - adding missing avatar (w/o coverPhoto)
  - adding missing coverPhoto (w/o avatar)
  - adding both missing assets
  - replacing existing avatar (w/o replacing coverPhoto)
  - replacing existing coverPhoto (w/o replacing avatar)
  - replacing both existing assets

**`createVideo` metadata cases to test (some can be combined):**
  - w/ both video and thumbnail
  - w/ video, but w/o thumbnail
  - w/ thumbnail, but w/o video
  - with valid/invalid license set
  - w/ publishedBeforeJoystream set
  - w/o any metadata and assets set


**Possible additional failure cases to include:**
- Invalid metadata when creating/updating a channel/video: expect exit code `400` - `InvalidInput`
- Invalid asset file paths in metadata when creating/updating channel/video
- Updating channel/adding videos/removing channel assets/removing channel with using non-owner context that matches selected account (multiple possible combinations): expect exit code `405` - `AccessDenied`
- Executing any of the commands with context not matching selected account: expect exit code `405` - `AccessDenied`
- Updating/removing a non-existing channel/video: expect exit code `400` - `InvalidInput`
- Removing channel assets that don't belong to the given channel: expect exit code `400` - `InvalidInput`

### **_TBC..._**

### Things to keep in mind
- Commands that are executed from different accounts or those that don't affect nonces in any way should be executed simultaneously, as this will save a huge amount of time. Some adjustments in CLI will be needed in order to support that (since currently the account is selected "globally" and cannot be changed on "per command" basis)

**Comment by dobertRowneySr** [comment]
# Basic Runtime Upgrade - single machine

## Chain state PRE-migration:
- A running joystream node with the Sumer runtime
- `channels.size > 0` Channels & `videos.size > 0` videos exist (possibly created in this flow)

## Chain state POST-migration:
- Operation WG 2 (BETA) & Operation WG 3 (GAMMA) exists and are correctly initialized
- `channels.size == 0` && `videos.size == 0` (i.e. there are no channels or videos immediately after migration)

## (Optimistic) Flow
- Start chain with Sumer runtime - no initial state
-  Add some content (without upload to storage node)
- Test Pre-migration Conditions
- Perform runtime upgrade to Giza
- Test Post-migration conditions
- Start Query-Node (starts indexing and processing after upgrade) - test handles initial state and bootstrap correctly

**Comment by Lezek123** [comment]
### Testing flows:

1. **Managing curator groups through Joystream CLI: https://github.com/Joystream/joystream/issues/2836**
    - **Requires** query node
    - **Requires** hired curator lead
    - **Tests** Joystream CLI (curator groups)
    - **Tests** Query node (curator groups)
    - **_Optionally_**: May create curator groups to be used by other tests

1. **Managing video/channel categories through Joystream CLI: https://github.com/Joystream/joystream/issues/2837**
    - **Requires** query node
    - **Requires** hired curator lead
    - **Tests** Joystream CLI (video/channel categories)
    - **Tests** Query node (video/channel categories)
    -  **_Optionally:_** May create channel/video categories to be used by other tests

1. **Managing channels and videos through CLI: https://github.com/Joystream/joystream/issues/2838**
    - **Requires** query node
    - **Requires** hired curator lead
    - **Requires** storage node setup to support channel bags (operator set, high enough bucket objects/size limits, metadata set, dynamic bag policy set etc.)
    - **Requires** distributor node setup to support channel bags (operator set, high enough bucket objects/size limits, metadata set, dynamic bag policy set etc.)
    - **_Optionally:_** Existing curator groups (would allow skipping the creation of those)
    - **_Optionally:_** Existing channel/video categories (would allow skipping the creation of those)
    - **Tests** Joystream CLI (videos and channels)
    - **Tests** Query node (videos and channels)
    - **Tests** Storage node (indirectly: uploading, providing assets to distributor node)
    - **Tests** Distributor node (fetching channel / video assets)

1. **Distributor node CLI:** https://github.com/Joystream/joystream/issues/2839
    - **Requires** query node
    - **Requires** hired distribution lead
    - **Tests** Distributor node (all CLI commands)
    - **Tests** Query node (distribution-related mappings)

1. **Reuploading assets through Joystream CLI:** https://github.com/Joystream/joystream/issues/2840
    - **Requires** query node
    - **Requires** hired storage lead
    - **Requires** storage node setup to support channel bags (operator set, high enough bucket objects/size limits, metadata set, dynamic bag policy set etc.)
    - **Requires** distributor node setup to support channel bags (operator set, high enough bucket objects/size limits, metadata set, dynamic bag policy set etc.)
    - **Tests** Joystream CLI (assets upload failure, `reuploadAssets` command)
    - **Tests** Storage node (indirectly: uploading, providing assets to distributor node)
    - **Tests** Query node (`StorageDataObject.accepted`)

1. **Distributor node assets fetching:** https://github.com/Joystream/joystream/issues/2841
    - **Requires** query node
    - **Requires** storage basic storage node setup (operator set, high enough bucket objects/size limits, metadata set, some bag(s) assigned)
    - **Requires** basic distributor node setup (operator set, high enough bucket objects/size limits, metadata set, some bag(s) assigned etc.)
    - **Tests** Storage node (indirectly: uploading, providing assets to distributor node)
    - **Tests** Distributor node (`/assets` endpoint - `GET` and `HEAD` responses and response headers depending on request headers and object state during request)

1. **Distributor node caching policy:** https://github.com/Joystream/joystream/issues/2842
    - **Requires** query node
    - **Requires** storage basic storage node setup (operator set, high enough bucket objects/size limits, metadata set, some bag(s) assigned)
    - **Requires** basic distributor node setup (operator set, high enough bucket objects/size limits, metadata set, some bag(s) assigned etc.)
    - **Tests** Storage node (indirectly: uploading, providing assets to distributor node)
    - **Tests** Distributor node (assets caching and cache eviction policy)

1. **Distributor node cache cleanup:** https://github.com/Joystream/joystream/issues/2843
    - **Requires** query node
    - **Requires** storage node setup to support channel bags (operator set, high enough bucket objects/size limits, metadata set, dynamic bag policy set etc.)
    - **Requires** distributor node setup to support channel bags (operator set, high enough bucket objects/size limits, metadata set, dynamic bag policy set etc.)
    - **Requires** hired distributor lead
    - **Tests** Storage node (indirectly: uploading, providing assets to distributor node)
    - **Tests** Distributor node (unsetting bag from bucket, cache cleanup functionality)

1. **Distributor node operator API buckets management:** https://github.com/Joystream/joystream/issues/2844
    - **Requires** query node
    - **Requires** storage basic storage node setup (operator set, high enough bucket objects/size limits, metadata set, some bag(s) assigned)
    - **Requires** basic distributor node setup (operator set, high enough bucket objects/size limits, metadata set, some bag(s) assigned etc.)
    - **Tests** Storage node (indirectly: uploading, providing assets to distributor node)
    - **Tests** Distributor node (setting buckets through authorized `operatorAPI`, `/buckets` endpoint, cache cleanup)

### Findings
1. Those flows (at least most of them) can't be executed in parallel. They require the storage/distributor node state (or related runtime state) not to be modified by other flows while they are still running, for example:
    - no items other than the expected ones should be evicted from distributor node cache during flows `3`, `5-9` (while flow `7` actually relies on cache eviction to be triggered)
    - we don't wan't bags we rely on in one flow to be unassigned from the distributor node in some other flow 
    - we don't want the storage node to become unreachable unless this is part of the test (which it is in case of `5`)
    - etc.

    The only solution that could make that (running in parallel) possible is for each test to use their own dedicated storage/distribution node and not modify the runtime state that would affect other storage/distributor nodes. That would make those tests much more demanding (resource-wise) and less flexible though.

2. Minor modifications in the CLI / Distributor node are needed
    - Joystream CLI: Allow passing the key as flag, which would make it possible to execute multiple commands in parallel
    - Distributor / Joystream CLI: Try to improve nonce management so that multiple commands from the same account can also be executed in parallel (if possible)
    - Joystream CLI: Normalize `ExitCode` when data objects cannot be uploaded (this now depends on whether there is any storage bucket available at all or not)
    - Distributor node: use `503` response when the data object cannot be retrieved from storage node (to differentiate it from generic `503` response)
    - Distributor node: Perhaps exposing more cache-related state could be useful for the tests (especially flow `7`)

3. Modifications in the testing framework are needed:
    - Support for typesafe graphql queries (can copy the setup from https://github.com/Joystream/joystream/tree/olympia)
    - Support to execute CLIs (already done in https://github.com/Lezek123/substrate-runtime-joystream/commit/1a6ad5e18b2873766dadb20b101a47f9047bbe35)
    - A simple way (fixture?) to generate and upload a data object of given size to a specified bag
    - A quick way (fixture / flow) to initialize a distributor node (and probably also storage node) with a worker, bucket, metadata etc. Doing it via CLI is not very optimal + it's better to separate it from CLI-specific flows. 

4. Running the flows above against a network (multiple storage / distribution nodes) should be possible. In case we want the flow to try to fetch objects from multiple distributor nodes we could adjust the flows (like flow `3` for example) to use multiple distributor node endpoints provided via `env` or fetched from the query-node. Other flows targetting specifically network tests (ie. checking which storage node the distributor node would use for feching the data, measuring the time etc.) could also be created.
